File: 2020/19_MonsterMessages/test_rule.py
# ======================================================================
# Monster Messages
#   Advent of Code 2020 Day 19 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                    t e s t _ r u l e s . p y
# ======================================================================
"Test single rule for Advent of Code 2020 day 19, Monster Messages"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import rule

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE_ZERO = '0: 4 1 5'
EXAMPLE_ONE = '1: 2 3 | 3 2'
EXAMPLE_FOUR = '4: "a"'

# ======================================================================
#                                                               TestRule
# ======================================================================


class TestRule(unittest.TestCase):  # pylint: disable=R0904
    "Test Rule object"

    def test_empty_init(self):
        "Test the default Rule creation"

        # 1. Create default Rule object
        myobj = rule.Rule()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, None)
        self.assertEqual(myobj.number, None)
        self.assertEqual(myobj.definition, None)

    def test_text_zero(self):
        "Test the Rule object creation from text with one alternate"

        # 1. Create Rule object from text
        myobj = rule.Rule(EXAMPLE_ZERO)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 8)
        self.assertEqual(myobj.number, 0)
        self.assertEqual(myobj.definition, '4 1 5')

        # 3. Check methods
        self.assertEqual(myobj.is_constant(), False)
        self.assertEqual(myobj.alternatives(), ['4 1 5'])

    def test_text_one(self):
        "Test the Rule object creation from text with two alternates"

        # 1. Create Rule object from text
        myobj = rule.Rule(EXAMPLE_ONE)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 12)
        self.assertEqual(myobj.number, 1)
        self.assertEqual(myobj.definition, '2 3 | 3 2')

        # 3. Check methods
        self.assertEqual(myobj.is_constant(), False)
        self.assertEqual(myobj.alternatives(), ['2 3 ', ' 3 2'])

    def test_text_four(self):
        "Test the Rule object creation from text with constant rule"

        # 1. Create Rule object from text
        myobj = rule.Rule(EXAMPLE_FOUR)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 6)
        self.assertEqual(myobj.number, 4)
        self.assertEqual(myobj.definition, '"a"')

        # 3. Check methods
        self.assertEqual(myobj.is_constant(), True)
        self.assertEqual(myobj.letter(), 'a')


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      t e s t _ r u l e . p y                   end
# ======================================================================

File: 2020/19_MonsterMessages/test_rule.lua
-- ======================================================================
-- Monster Messages
--   Advent of Code 2020 Day 19 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ r u l e . l u a
-- ======================================================================
-- Test Rule for Advent of Code 2020 day 19, Monster Messages

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local rule = require('rule')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = "1: 2 3 | 3 2"
local EXAMPLE_TEXT4 = '4: "a"'

-- ======================================================================
--                                                               TestRule
-- ======================================================================

function test_empty_init()
  -- Test the default Rule creation

  -- 1. Create default Rule object
  local myobj = rule:Rule()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(myobj.number, nil)
  luaunit.assertEquals(myobj.definition, nil)
  

end

function test_text_init()
  -- Test the Rule object creation from text

  -- 1. Create Rules object from text
  local myobj = rule:Rule({text=EXAMPLE_TEXT})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.text, EXAMPLE_TEXT)
  luaunit.assertEquals(myobj.number, "1")
  luaunit.assertEquals(myobj.definition, "2 3 | 3 2")
  
  -- 3. Test methods
  luaunit.assertEquals(myobj:is_constant(), false)
  luaunit.assertEquals(myobj:alternatives(), {"2 3", "3 2"})
end

function test_rule4()
  -- Test the Rule object creation from text

  -- 1. Create Rules object from text
  local myobj = rule:Rule({text=EXAMPLE_TEXT4})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.text, EXAMPLE_TEXT4)
  luaunit.assertEquals(myobj.number, "4")
  luaunit.assertEquals(myobj.definition, '"a"')
  
  -- 3. Test methods
  luaunit.assertEquals(myobj:is_constant(), true)
  luaunit.assertEquals(myobj:letter(), "a")
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                 t e s t _ r u l e . p y                end
-- ======================================================================
